proposals/0530-strict-box-denesting.rst

```diff
@@ -0,0 +1,527 @@
+
+
+
+::
+
+   -- isomorphic to Data.Constraints.Dict (in constraints)
+   data Example5 (c :: Constraint) where
+     MkExample5 :: c => Example5 c
+
+The ``MkExample5`` constructor of ``Example5 c`` has one field, which holds the
+run-time evidence for the constraint ``c``, represented by a GC-managed
+pointer, usually to an instance-method-dictionary-object.  However,
+even with ``-fdicts-strict``, GHC will not make this a strict field, as
+doing so is incompatible with the newtype-class optimization.
+Today, the only evidence fields GHC makes strict are the implicit
+equality evidence fields of a GADT constructor.  So, ``MkExample5``
+does not satisfy condition 4 and is therefore not denestable.
+
+..
+  TODO: Even more examples? There are quite a few in the Motivation
+  section as well.
+
+  Do I really need a DataToTag example? I hypothesize nobody is
+  actually affected by that change except the proposal implementors.
+
+
+Effect and Interactions
+-----------------------
+
+The main effect of this proposal would be to cause some Haskell
+programs to run faster and use less memory.
+
+This proposal interacts nontrivially with the ``dataToTag#`` primop,
+which currently always evaluates its argument and returns the index of
+the data constructor (if it exists) used to produce the resulting
+evaluated heap object.  Since ``-fstrict-box-denesting`` affects the
+set of things that produce new evaluated heap objects, it can affect
+the result of a call to the current implementation of
+``dataToTag#``.  For example, ``dataToTag# (MkExample1 [True])`` with
+its current implementation may return either
+
+* the index ``0#`` of the ``MkExample1`` constructor, or
+* the index ``1#`` of the ``(:)`` constructor of the underlying
+  heap object ``(:) True []``
+
+\...depending on whether or not the ``-fstrict-box-denesting`` flag is
+active when in the module containing ``Example1`` is compiled.  This
+proposal suggests to retain the current behavior of ``dataToTag#``
+regardless of the ``-fstrict-box-denesting`` flag by allowing a
+different implementation to be used for types with a denestable
+constructor, via run-time evidence of a ``DataToTag`` typeclass
+constraint.
+
+
+
+Costs and Drawbacks
+-------------------
+1. Today, most or all of GHC's boxed primitive types such as
+   ``ByteArray#`` cannot be safely stored in a value with lifted type,
+   as they are zero-tagged pointers (so they will be entered when
+   evaluated) but entering them results in a panic.  These panics have
+   been useful for GHC developers in the past, but are incompatible
+   with this proposal as it is currently written, since it allows
+   a wrapper such as ``ShortByteString`` to be represented at runtime
+   by its underlying ``ByteArray#``.
+
+   Since entering these objects involves an indirection that would
+   largely defeat the purpose of denesting their lifted wrappers, this
+   must be dealt with by changing the proper tag for their pointers.
```

Comment:
"These objects" refers to "GHC's boxed primitive types" from the previous paragraph: I want to change the proper tag of `ByteArray#` to 1.

Reply:
I still don't understand completely. You make it sound like the *tag* is the issue. I'm sure it is not. Specifically, I'm getting hung up on

>  Since entering these primitive objects involves **an indirection** that would **largely defeat the purpose of denesting** their lifted wrappers, this must be dealt with by changing the proper tag for their pointers.

What indirection? Do you mean the call to the enter code or do you mean the residual `IND` heap object after evaluation of the thunk concluded? I don't see how the latter defeats the purpose of denesting. Lifted Newtypes have `IND`, too; the indirection eliminated there is a different one, namely that of the *additional* data constructor incurred by the, e.g., `SBS` data constructor.

I still don't understand how you can deal with the `IND` indirection by changing poitner tags.

Perhaps you didn't mean `IND` indirections after all? Then I'm at a loss, because that sentence doesn't make sense for "indirection = entering the heap object" to me.

---

In particular, I don't see how giving `ByteArray#` a different pointer tag makes a difference to the enter code problem. Consider

```hs
let sbs = ... :: ShortByteString
in ... f sbs + g sbs ...
```

What tag does `sbs` have prior to being evaluated? I say it should be 0, it's a thunk after all. That's all fine, until `sbs` is evaluated in `f`. Now the original pointer `sbs` still has tag 0 but we have `sbs -> IND -> ByteArray# ...` in the heap. When `g` evaluates `sbs`, it enters it due to the lack of a tag. That will execute the enter code for `IND` and that in turn executes the enter code of `ByteArray#`. If that enter code is still a panic, we'll panic.

IMO the panic is a clear violation of our "ABI", because it should always be safe to enter a heap object (to get its tag, for example). It's also different to what unlifted data types do.

Reply:
> What indirection? Do you mean the call to the enter code or do you mean the residual IND heap object after evaluation of the thunk concluded?

I mean calling the enter code.



> In particular, I don't see how giving `ByteArray#` a different pointer tag makes a difference to the enter code problem. Consider
> 
> ```haskell
> let sbs = ... :: ShortByteString
> in ... f sbs + g sbs ...
> ```
> 
> What tag does `sbs` have prior to being evaluated? I say it should be 0, it's a thunk after all. That's all fine, until `sbs` is evaluated in `f`. Now the original pointer `sbs` still has tag 0 but we have `sbs -> IND -> ByteArray# ...` in the heap. When `g` evaluates `sbs`, it enters it due to the lack of a tag. That will execute the enter code for `IND` and that in turn executes the enter code of `ByteArray#`. If that enter code is still a panic, we'll panic.

Changing the pointer tag for `ByteArray#` does solve this problem. When `sbs` is evaluated in `f` we end up with `sbs -> IND -> ByteArray#` in the heap, but the pointer to a `ByteArray#` inside the `IND` is (or should be) correctly tagged and does not need to be entered!

Reply:
It looks like according to [the ghc wiki](https://gitlab.haskell.org/ghc/ghc/-/wikis/commentary/rts/storage/heap-objects?version_id=a98bed4769d81f169a2470dc6f0f5fabe0a0f54c) "The entry code for all indirections simply enters the closure it points to." And I suppose that makes sense: the evaluated indirectee may itself be a function with some non-trivial entry code. (Although this seems obviously suboptimal for the extremely common case of updatable thunks returning ordinary data with arity zero... I wonder about the potential performance benefit of adding and using a separate `DATA_IND` closure type that doesn't enter its indirectee...)

But even if for whatever reason `ByteArray#` and friends **must** be made enterable, entering an object still involves:

* An indirection/one level of pointer-chasing to get the entry code location
* pushing a continuation to the stack for the entry code to return to
* a dynamic/indirect jump to run the entry code
* a dynamic/indirect jump back from the entry code

...which is not great, even compared to the just status quo of one heap lookup to access a field of a heap data object. So for performance reasons it is very desirable that if `ByteArray#` can be passed around in place of lifted values, they do so with some tag other than zero.